### PR TITLE
add missing "revision" parameter to addModel step.

### LINKED ISSUE #1042: Pinned revision in bundle not taken into account when deploying with model.deploy()
From the conversation at: https://matrix.to/#/!xzmWHtGpPfVCXKivIh:ubuntu.com/$gVXe8o8uyDewViyBtlQi0QBFn6eWfy4nGqSbegxg0jg?via=ubuntu.com&via=matrix.org&via=fsfe.org

Hi! 

When trying to deploy this bundle : https://github.com/canonical/iam-bundle/blob/main/bundle.yaml.j2, I see that the revision for some of the charms are pinned (e.g: rev52 for self-signed-certificates). But when I deploy the bundle with this code :
```
identity_platform = await model.deploy(
    'identity-platform',
     channel='edge',
     trust=True,
)
```
The pinned revision didn't get taken into account and instead I get the latest revision of the charms (rev127 in this case)
```
self-signed-certificates                      active       1  self-signed-certificates             latest/edge    127  10.152.183.67   no 
```
This doesn't happen when I do `juju deploy identity-platform --channel=latest/edge --trust`

### Environment
```
(.env) ubuntu@ubuntu:~$ juju --version
3.1.7-genericlinux-amd64

(.env) ubuntu@ubuntu:~$ sudo microk8s version
MicroK8s v1.28.7 revision 6532

(.env) ubuntu@ubuntu:~$ python3 --version
Python 3.10.12

(.env) ubuntu@ubuntu:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.4 LTS
Release:	22.04
Codename:	jammy

(.env) ubuntu@ubuntu:~$ juju controllers
Use --refresh option with this command to see the latest information.

Controller   Model      User   Access     Cloud/Region        Models  Nodes  HA  Version
controller*  pylibjuju  admin  superuser  microk8s/localhost       2      1   -  3.1.7  

(.env) ubuntu@ubuntu:~$ pip3 show juju 
Name: juju
Version: 3.4.0.0
Summary: Python library for Juju
Home-page: https://github.com/juju/python-libjuju
Author: 
Author-email: 
License: Apache 2
Location: /home/ubuntu/.env/lib/python3.10/site-packages
Requires: hvac, kubernetes, macaroonbakery, packaging, paramiko, pyasn1, pyRFC3339, pyyaml, toposort, typing-inspect, websockets
Required-by: 
```

### PR DESCRIPTION
**Pull request by Thanhphan1147**
#### Description

Fixes #1042.

When deploying a bundle, although the plan is properly generated with the correct revision for the charms in the bundle, and the change steps are generated with the correct parameter. I believe that the `addCharm` step does not take into account the specified "revision" of the charm to deploy, as a result the revision deployed is the latest revision instead of the specified revision. ref: https://github.com/juju/python-libjuju/blob/5ed5ae461514feb84dc0f96c2e46b6fab9f35861/juju/bundle.py#L709-L713

This PR simply adds the missing parameter to the `AddCharmChange` step and propagate the value to `client.CharmOrigin` to add the correct revision.

#### QA Steps
1. Create a bundle containing one or more charms with revision
2. Deploy the bundle with `model.deploy`
3. The deployed charms in the bundle should have the revision specified in the bundle.

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- ~~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed package~~ (I think this is too small of a change to warrant a doc update)

### CODE REVIEW

File: juju/bundle.py
Comment (jameinel) [inline juju/bundle.py:662]: I'm a little surprised that this is all it takes, though I guess the issue with the code that just follows this is that the `from_dict` base class looks for all the listed items and then will look to set the attributes based on the names that are given.

**Review by jameinel — COMMENTED** [review]
It definitely feels like we are missing a test of some form (plausibly an integration test), to ensure that bundles with charm revisions are properly deployed, but the change itself seems plausible.

File: juju/bundle.py
Comment (Thanhphan1147) [reply juju/bundle.py:662]: I'm not that knowledgeable about the internals of python-libjuju but when debugging I saw that the revision was passed to this step as a parameter so to me it makes sense from a high-level view. also the other parameters like series and channel are also done in a similar manner.

It seems like I also need to update the unit tests, I'll find the time to work on this. I'll also add an integration test

Comment (Thanhphan1147) [reply juju/bundle.py:662]: I fixed the issues with unit tests and also added an integration test for the pinned revision

**Review by cderici — CHANGES_REQUESTED** [review]
Thanks for the change @Thanhphan1147! A little fix for [the failing test](https://github.com/juju/python-libjuju/actions/runs/8708688290/job/23889160063?pr=1043#step:6:512) is needed on this one, let's get that passing, then we can take over and do some manual QA to get this landed👍

File: tests/integration/test_model.py
Comment (Thanhphan1147) [reply tests/integration/test_model.py:189]: Thanks for the suggestion, I've committed the changes, I will also test it out locally to see if everything is okay and make additional changes if needed

**Comment by Thanhphan1147** [comment]
the added integration test seems to be passing locally for me.

**Comment by Thanhphan1147** [comment]
Seems like a few other integration tests are having issues, maybe a few updates on the bundle definitions as the new charms revision no longer supports old series.

The one I added seems to be fine: 
```
[gw0] [ 45%] PASSED tests/integration/test_model.py::test_deploy_bundle_with_pinned_charm_revision 
```

**Review by cderici — APPROVED** [review]
This LGTM, manual QA performed against 3.4, everything seems to be working fine 👍 

@Thanhphan1147 I can confirm that the failing tests are known test failures that'll be fixed soon (and unrelated to this change), this can land safely 👍thanks again for the contribution, much appreciated!

**Comment by jameinel** [comment]
@Thanhphan1147 Caner is off until Monday. I would tend to wait for him to merge this, but if you need it sooner, let us know.

**Comment by Thanhphan1147** [comment]
@jameinel nothing urgent from me, thanks for your help on this PR!

**Pull request merged**
